Clarify Send Order (#417)
moq-wg/moq-transport: draft-ietf-moq-transport.md

- Prioritization Schemes.
- Congestion Algorithms and impacts.
- Considerations for merging multiple streams across domains onto single
  connection and interactions with specific prioritization schemes.

## Object Priority {#object-priority}

Goals of Object Priority include:
* Enable objects to be delivered in the order intended by the publisher
* Allow nodes and relays to delay or skip some objects under congestion
* Create more reliable behavior of relays

The publisher's Object Priority SHOULD be followed when possible. The

> contents of each object are still delivered in order.

Reviewer: s/still/always/ ?

> contents of each object are still delivered in order.

Author: ```suggestion
contents of each object are delivered in order.
```

- Prioritization Schemes.
- Congestion Algorithms and impacts.
- Considerations for merging multiple streams across domains onto single
  connection and interactions with specific prioritization schemes.

## Object Send Order {#send-order}

Goals of Send Order include:

> * Enable objects to be delivered in the order intended by the publisher

Reviewer 2: This sounds right for subscribe, I think it would be reasonable for Fetch to add an preference to deliver in order, or in reverse order.

between objects.

A sender MUST send each object over a dedicated stream.  The library
should support prioritization ({{priority-congestion}}) such that
streams are transmitted in send order.
## Object Send Order {#send-order}

A receiver MUST NOT assume that objects will be received in send order,
for the following reasons:
Goals of Send Order include:
* Enable objects to be delivered in the order intended by the publisher
* Allow nodes and relays to delay or skip some objects under congestion
* Create more reliable behavior of relays

The publisher's send order SHOULD be followed when possible. The contents
of each object are still delivered in order.

A receiver MUST NOT assume that objects will be received in send order,
for reasons including:
* Newly encoded objects can have a smaller send order than outstanding
  objects.
* Packet loss or flow control can delay the send of individual streams.
* Packet loss or flow control can delay sending individual streams.
* The sender might not support stream prioritization.

TODO: Refer to Congestion Response and Prioritization Section for
further details on various proposals.

> ### Proposal - Ordering by Priorities {#ordering-by-priorities}

Reviewer 3: I feel like this proposal has been deprecated since implementing stream mapping, since the priority is in the stream header now. Relays are expected to respect it and forward streams as they were received.

Reviewer 4: Priority is saved as part of "Canonical Object" along with Object Forwarding preference . So that when they are served from  cache, regardless of stream mapping, right priority is applied and mapped to the right stream.

> When nodes or relays have to choose which object to send next, they

Reviewer 3: I think this text could be salvaged by recommending how `send_order` is computed. Otherwise it just conflicts with the above paragraphs that just tell you to use send order.

```suggestion
When the publisher chooses which object to send next via `send_order`, they
```

> * Within the same group, objects with a lower priority number (e.g. 1)

Reviewer 3: Priorities are applied to streams so this gets super confusing based on the stream mode. The object_id is not always known after the stream header (track/group mode); am I supposed to change the stream priority in the middle of a stream?

Reviewer 4: @kixelated can you please elaborate on what needs clarification here ?   Streaming modes naturally fit with priorities and we have implemented the same.

If stream per group is selected , all the objects with in a group have same priority.  So objects are delivered in the receive order at a relay within a stream.
if stream per object is selected and if the application marks all the objects with same priority within a group, then objects end up on the same priority queue level and are sent in the receive order.  If application marks each object with a different priorities, they end up on different levels in the queue and are sent according to the priority order.

The above text seems to cover all the cases if am reading it right.

Reviewer 3: So in general this text needs some clarification on if it's normative or non-normative. I interpret it as normative, outlining the algorithm that each publisher uses to compute `send_order`.

If so, it breaks some use-cases, like using send order for head-of-line blocking. It also just makes the `send_order` redundant, as a publisher could compute the send order using `group_id`, `object_id` and these rules. 

And what happens when there's a mismatch? Is a relay supposed to error if `send_order` does not follow these rules?

I mean this was the crux of the debate way back then and why there are two sections. I thought the publisher should bake the rules into the send_order, while Christian thought we should have explicit rules. This PR is effectively doing both by merging the sections together; allowing publishers to choose the send order per stream but dictating the rules?

Reviewer: We've got text elsewhere that says that:

* when using stream-per-group, all objects have the same priority (sort of, see note)
* when using stream-per-group, the relay can't skip or reorder objects within the group

It might help clarify if this section referenced those bits, or explained the intersection of forwarding preferences and priorities.

Note: Our wire format allows the same group to be transmitted with different priorities if the group is carried on two or more streams (eg: disconnect in the middle of a group, objects 1-10 on stream 1, objects 11-20 on a different stream, or connection).

> * Within the same group and priority, objects with a lower object-id

Reviewer 3: Regardless of send order?

Reviewer: Here priority == send order, so the qualifier (within the same group and priority) is that these objects have identical send order.

> * Objects from later groups are sent before objects of previous groups.

Reviewer 3: What happens when a newer group has a higher send order?

I think it's a great recommendation to send newer groups first... unless you want head-of-line blocking for reliable live or VOD. This text makes it seem illegal?

Reviewer: I think this text should be scoped to tiebreakers:

"When objects from two or more groups have the same priority, send the objects from the more recent group first"

But for the "reliable live" case, you probably still want a subscriber side override to get groups in ascending order.

group.

* Object Send Order: An integer indicating the object send order

> {{send-order}}.

Reviewer 2: So the algorithm above refers to "priority". I agree we only need one integer but need to align names.

Author: Good point, personally I think I like "Object Priority" based on how it's discussed and used.  But suggestions are welcome.
